### PR TITLE
Add ClassifierGLM

### PR DESCRIPTION
**Pull request by BalzaniEdoardo**
# Summary

Adds ~~`CategoricalGLM` and `CategoricalPopulationGLM`~~ `ClassifierGLM` and `ClassifierPopulationGLM` for modeling categorical outcomes such as behavioral choices.

## New Classes

- **~~`CategoricalGLM`~~** **`ClassifierGLM`**: Fit classification models for single subjects
- **~~`CategoricalPopulationGLM`~~** **`ClassifierPopulationGLM`**: Fit multiple subjects in parallel

## API

Both classes follow the familiar GLM API with key differences:

| Method | Description |
|--------|-------------|
| `fit(X, y)` | Fit model where `y` contains integer category labels |
| `predict(X)` | Returns predicted category labels (not rates) |
| `predict_proba(X)` | Returns probabilities or log-probabilities for each category |
| `simulate(key, X)` | Generate synthetic categorical data |
| `score(X, y)` | Compute mean log-likelihood |
| `set_classes(y)`| Set the classes labels (default `[0,...,N-1]`) | 

### Parameter shapes

| Model | `coef_` | `intercept_` |
|-------|---------|--------------|
| ~~`CategoricalGLM`~~ | ~~`(n_features, n_categories - 1)`~~ | ~~`(n_categories - 1,)`~~|
| ~~`CategoricalPopulationGLM`~~ | ~~`(n_features, n_neurons, n_categories - 1)`~~| ~~`(n_neurons, n_categories - 1)`~~|
| `ClassifierGLM` | `(n_features, n_categories )` | `(n_categories,)`|
| `ClassifierPopulationGLM` | `(n_features, n_neurons, n_categories)` | `(n_neurons, n_categories)` |

## Implementation Details

- New validators: `ClassifierGLMValidator` and `PopulationClassifierGLMValidator` in `validation.py`
- Uses `log_softmax` as the default inverse link function
- Compatible with all existing regularizers (Ridge, Lasso, ElasticNet, GroupLasso)
- Integrated with pynapple for time series support
- Full scikit-learn compatibility for cross-validation and pipelines

## Documentation

- New how-to guide: `docs/how_to_guide/glm_for_classification.md`
- Added "Categorical GLM" section to quickstart
- Added to API reference

### CODE REVIEW

**Comment by sjvenditto** [comment]
A couple comments:
- Should we rename the class to something like `ClassifierGLM` or `ClassificationGLM`, and use parameters like `n_classes` (basically classes instead of categories)? I feel like the term "categorical" usually refers to the feature type or the distribution type, where the model itself is associated with classification. The parameter naming would also be more consistent with sklearn's `LogisticRegression`
- Why not use `n_categories` coefficients like sklearn? Personally, I feel like this is more useful because I would want to compare the coefficients of all the categories, and I'd want the model fit to give that to me instead of me having to extract the "implicit" category myself. Also, the per-category coefficients can be meaningful, e.g. this sklearn example with MNIST classification: https://scikit-learn.org/stable/auto_examples/linear_model/plot_sparse_logistic_regression_mnist.html
- If we really want to go the full sklearn route, we may also consider allowing other data types, e.g. strings, that we convert internally to integer labels. but `simulate` and `predict` would return the original label types, and we'd have the attribute `classes_` to list out the extracted class labels

**Comment by BalzaniEdoardo** [comment]
> A couple comments:
> 
> * Should we rename the class to something like `ClassifierGLM` or `ClassificationGLM`, and use parameters like `n_classes` (basically classes instead of categories)? I feel like the term "categorical" usually refers to the feature type or the distribution type, where the model itself is associated with classification. The parameter naming would also be more consistent with sklearn's `LogisticRegression`

I agree with all these points, I'll make the changes tomorrow.

> * Why not use `n_categories` coefficients like sklearn? Personally, I feel like this is more useful because I would want to compare the coefficients of all the categories, and I'd want the model fit to give that to me instead of me having to extract the "implicit" category myself. Also, the per-category coefficients can be meaningful, e.g. this sklearn example with MNIST classification: https://scikit-learn.org/stable/auto_examples/linear_model/plot_sparse_logistic_regression_mnist.html

This doesn't play well with the unregularized version because to identify the model you need some sort of regulation. Otherwise you'll have K params but only K-1 degrees of freedom. 

> * If we really want to go the full sklearn route, we may also consider allowing other data types, e.g. strings, that we convert internally to integer labels. but `simulate` and `predict` would return the original label types, and we'd have the attribute `classes_` to list out the extracted class labels

That could be, not too hard to do that.

**Comment by BalzaniEdoardo** [comment]
ok, for @sjvenditto point num 2, we can change the `from_model_param` function to convert from k-1 weghts to k weights, same thing with the `to_model_params` to do the reverse

**Review by gviejo — APPROVED** [review]

File: src/nemos/glm/validation.py [hunk 1]
```diff
@@ -372,3 +387,272 @@ def get_empty_params(self, X, y) -> GLMParams:
...
+
+    Validates and transforms user-provided parameters, inputs, and checks consistency
+    between parameters and data for classifier GLMs. Classifier GLMs have:
+    - 2D coefficients: shape (n_features, n_classes - 1) or dict of (n_features, n_classes - 1) arrays
+    - 1D intercept: shape (n_classes - 1,)
+    - 2D input X: shape (n_samples, n_features) or pytree of same
+    - 1D output y: shape (n_samples,) containing integer class labels
+    """
+
+    extra_params: ClassifierExtraParams = field(kw_only=True)
```
Comment (sjvenditto) [inline src/nemos/glm/validation.py]: why does this need to be a class? why can't it just be a dictionary defined here?

File: src/nemos/glm/validation.py [hunk 2]
```diff
@@ -372,3 +387,272 @@ def get_empty_params(self, X, y) -> GLMParams:
...
+
+
+@dataclass(frozen=True, repr=False)
+class ClassifierGLMValidator(GLMValidator):
+    """
+    Validator for classifier GLM models.
+
+    Validates and transforms user-provided parameters, inputs, and checks consistency
+    between parameters and data for classifier GLMs. Classifier GLMs have:
+    - 2D coefficients: shape (n_features, n_classes - 1) or dict of (n_features, n_classes - 1) arrays
```
Comment (sjvenditto) [inline src/nemos/glm/validation.py]: should this be `n_classes`? not `n_classes-1`?

File: src/nemos/glm/validation.py [hunk 3]
```diff
@@ -372,3 +387,272 @@ def get_empty_params(self, X, y) -> GLMParams:
...
+
+
+@dataclass(frozen=True, repr=False)
+class PopulationClassifierGLMValidator(ClassifierGLMValidator):
+    """
+    Validator for population (multi-neuron) classifier GLM models.
+
+    Validates and transforms user-provided parameters, inputs, and checks consistency
+    between parameters and data for population classifier GLMs. Population classifier GLMs have:
+    - 3D coefficients: shape (n_features, n_neurons, n_classes - 1) or dict of same
```
Comment (sjvenditto) [inline src/nemos/glm/validation.py]: should this also be `n_classes`?

File: docs/how_to_guide/glm_for_classification.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    jupytext_version: 1.18.1
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+# Fit GLM for Classification
+
+The [`ClassifierGLM`](nemos.glm.ClassifierGLM) models categorical outcomes such as behavioral choices.
```
Comment (sjvenditto) [inline docs/how_to_guide/glm_for_classification.md]: I think we need to emphasize more strongly that the "Categorical" observation model is the only one compatible with this GLM (or "Categorical" + "Bernoulli" if we go that route). Alongside this description, we can have a bullet point in the key differences from GLM highlighting it, e.g. that observation model is limited / should not be set

File: docs/quickstart.md [hunk 5]
```diff
@@ -141,6 +141,57 @@ print(f"Model intercept shape: {population_model.intercept_.shape}")
...
+
+import nemos as nmo
+
+# Instantiate a categorical model for 3 categories
+n_classes = 3
+model = nmo.glm.ClassifierGLM(n_classes)
+
+```
+
+The `ClassifierGLM` follows a similar API to the standard GLM, with a few key differences:
```
Comment (sjvenditto) [inline docs/quickstart.md]: Same comment as above: I think we need to emphasize more strongly that the "Categorical" observation model is the only one compatible with this GLM (or "Categorical" + "Bernoulli" if we go that route). Alongside this description, we can have a bullet point in the key differences from GLM highlighting it, e.g. that observation model is limited / should not be set

File: docs/how_to_guide/glm_for_classification.md (hunk 4 above)
Comment (billbrod) [reply docs/how_to_guide/glm_for_classification.md]: that last bullet point should probably revolve around the fact that the ClassifierGLM does not have an observation model parameter in its init method

File: docs/how_to_guide/glm_for_classification.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+**Key differences from standard GLM:**
+- [`predict`](nemos.glm.ClassifierGLM.predict) returns predicted category labels
+- [`predict_proba`](nemos.glm.ClassifierGLM.predict_proba) returns (log-)probabilities for each category
+- [`set_classes`](nemos.glm.ClassifierGLM.set_classes) set class labels. By Default, classes are assumed to be `{0, ..., n_classes - 1}` however, this assumption can be overridden by calling `set_classes` passing an array of class labels.
+
+## Generate Synthetic Data
+
```
Comment (billbrod) [inline docs/how_to_guide/glm_for_classification.md:25]: I think there should be some text here highlighting that this is simulated (and with real data, this section would be replaced by loading in the data), because that might be easy to miss.

File: docs/quickstart.md [hunk 7]
```diff
@@ -141,6 +141,57 @@ print(f"Model intercept shape: {population_model.intercept_.shape}")
 
 ```
 
+### **Categorical GLM**
```
Comment (billbrod) [inline docs/quickstart.md]: should this section be called classifier glm to be consistent with the class name?

File: docs/quickstart.md [hunk 8]
```diff
@@ -141,6 +141,57 @@ print(f"Model intercept shape: {population_model.intercept_.shape}")
...
+
+# Instantiate a categorical model for 3 categories
+n_classes = 3
+model = nmo.glm.ClassifierGLM(n_classes)
+
+```
+
+The `ClassifierGLM` follows a similar API to the standard GLM, with a few key differences:
+
+- `predict` returns predicted category labels (not rates)
```
Comment (billbrod) [inline docs/quickstart.md]:
```suggestion
- `predict` returns predicted category labels (unlike the standard `GLM` model, which returns predicted firing rates).
```

with link to the GLM's predict method probably

File: docs/quickstart.md [hunk 9]
```diff
@@ -141,6 +141,57 @@ print(f"Model intercept shape: {population_model.intercept_.shape}")
...
+- `predict_proba` returns the probability of each category
+
+```{code-cell}
+
+import numpy as np
+num_samples, num_features = 100, 3
+
+# Generate a design matrix
+X = np.random.normal(size=(num_samples, num_features))
+# generate categorical observations (values 0, 1, or 2)
```
Comment (billbrod) [inline docs/quickstart.md]:
```suggestion
# simulate categorical observations (values 0, 1, or 2)
```

File: src/nemos/glm/glm.py [hunk 10]
```diff
@@ -227,6 +249,33 @@ def __init__(
         self.aux_ = None
         self.optim_info_ = None
 
+    @classmethod
+    def _validate_observation_class(cls, observation: obs.Observations):
```
Comment (billbrod) [inline src/nemos/glm/glm.py:250]: I feel like this GLM's validate method should assume the model's non-categorical, and the subclass should check for categorical ones

File: src/nemos/glm/glm.py [hunk 11]
```diff
@@ -1461,3 +1496,1015 @@ def __sklearn_clone__(self) -> PopulationGLM:
...
+            i.e. ``len(np.unique(y)) == n_classes``.
+
+        Notes
+        -----
+        :meth:`fit` and :meth:`initialize_solver_and_state` call ``set_classes`` internally,
+        making sure that the ``classes_`` attribute matches the provided input.
+        If you are fitting in batches by calling :meth:`update`, make sure that the ``classes_``
+        are correctly set by calling ``set_classes`` before starting the :meth:`update` loop.
+
+        Examples
```
Comment (billbrod) [inline src/nemos/glm/glm.py]: should probably show this being used with update, which is the only reason a user would call it, right?

Comment (billbrod) [reply src/nemos/glm/glm.py]: and probably some of the failures

File: src/nemos/glm/glm.py [hunk 12]
```diff
@@ -1461,3 +1496,1015 @@ def __sklearn_clone__(self) -> PopulationGLM:
...
+    # observation model inferred
+    _invalid_observation_types = ()
+
+    def _check_classes_is_set(self, method_name, y=None):
+        if self._classes_ is None:
+            raise RuntimeError(
+                f"Classes are not set. Must call ``set_classes`` before calling ``{method_name}``."
+            )
+
+    def set_classes(self, y: ArrayLike) -> ClassifierMixin:
```
Comment (billbrod) [inline src/nemos/glm/glm.py]: should document the errors raised

File: src/nemos/glm/glm.py [hunk 13]
```diff
@@ -1461,3 +1496,1015 @@ def __sklearn_clone__(self) -> PopulationGLM:
...
+        params :
+            The current model parameters, typically a tuple of coefficients and intercepts.
+        opt_state :
+            The current state of the optimizer, encapsulating information necessary for the
+            optimization algorithm to continue from the current state.
+        X :
+            The predictors used in the model fitting process. Shape ``(n_time_bins, n_features)``
+            or a ``FeaturePytree``.
+        y :
+            Class labels as integers, array of shape ``(n_time_bins,)`` for single neuron
```
Comment (billbrod) [inline src/nemos/glm/glm.py]: I think this is incorrect -- we allow any dtype as labels right? so not just float but also strings? (that's what the call to _encode_labels is doing, yeah)?

File: src/nemos/glm/glm.py [hunk 14]
```diff
@@ -1461,3 +1496,1015 @@ def __sklearn_clone__(self) -> PopulationGLM:
...
+            observation_model=observation_model,
+            inverse_link_function=inverse_link_function,
+            regularizer=regularizer,
+            regularizer_strength=regularizer_strength,
+            solver_name=solver_name,
+            solver_kwargs=solver_kwargs,
+        )
+        self._classes_ = None
+        self._class_to_index_ = None
+        self._skip_encoding = True  # default: assume labels are [0, n_classes-1]
```
Comment (billbrod) [inline src/nemos/glm/glm.py]: this feels like a dangerous default -- shouldn't we default to being safer at the expense of performance?

File: src/nemos/glm/validation.py [hunk 15]
```diff
@@ -206,46 +214,53 @@ def validate_consistency(
...
-        it to JAX arrays with the specified data type. Subclasses can extend
+        it to JAX arrays with the specified data type. Handles FeaturePytree
+        inputs by extracting the underlying data. Subclasses can extend
         this to add parameter-specific validation (e.g., checking that mask
         shape matches parameter dimensions).
 
         Parameters
         ----------
-        feature_mask : dict[str, jnp.ndarray] or jnp.ndarray
+        feature_mask : FeaturePytree, dict[str, jnp.ndarray], jnp.ndarray, or None
```
Comment (billbrod) [inline src/nemos/glm/validation.py]: isn't annotating the parameters here redundant with the type hinting?

File: src/nemos/glm/glm.py (hunk 14 above)
Comment (BalzaniEdoardo) [reply src/nemos/glm/glm.py]: switched to false

File: src/nemos/glm/glm.py (hunk 13 above)
Comment (BalzaniEdoardo) [reply src/nemos/glm/glm.py]: yep, i changed the description
